dev-docs/design/how-lit-html-works.md

```diff
@@ -0,0 +1,431 @@
+
+`TemplatePart`s are found by walking the tree of nodes in the `<template>` element and finding the marker annotations. If a marker is found, either in an attribute, as a marker comment node, or in text content, a `TemplatePart` is created and stored in the `parts` class field.
+
+When traversing the tree of nodes, the following work is done for the different Node types:
+
+##### Element
+
+For each attribute on the element, if it has a `$lit$` suffix, then the expression represents either a `PropertyPart`, `BooleanAttributePart`, `EventPart`, or `AttributePart`. These can be distinguished by the first character in the attribute name ([see Appendix Parts table](#parts)).
+
+An attribute on the element may also represent an `ElementPart` in the case where the attribute name starts with the marker sentinel value. E.g. `<div ${} ${}>` is marked up as `<div lit$random$0="" lit$random$1="">`.
+
+For each of the bound attributes identified on the element, they are removed from the element and a `TemplatePart` is created to record the location, name, and type of dynamic binding.
+
+##### Comment
+
+A Comment is usually either an expression marker, or a user-written comment with no expression associated. Occasionally though a user may have written an expression inside a comment. We might see a comment like `<!--<div>${text}</div>-->` in the template text, so we must scan comments for marker text.
+
+A comment that matches a marker annotation is a `TemplatePart` in ChildPart position.
+
+##### Text
+
+Usually text-position markers will be comment nodes, but inside `<style>`, `<script>`, `<textarea>`, `<title>` tags the markup that _looks_ like a comment (`<!-- lit -->`) will be inserted as _text_. So we must search for the marker within Text nodes. If found we split the Text node and insert a comment marker, and store a `TemplatePart` in ChildPart position referencing the comment marker's node index.
+
+### 2.ii. Create
+
+The create phase is performed only when rendering for **the first time to a specific Part**.
+
+In the `lit-html` source code this phase is the [instantiating and cloning](https://github.com/lit/lit/blob/64fb960246e8f1eae982c2f09fca9759001756af/packages/lit-html/src/lit-html.ts#L1534-L1535) of a `TemplateInstance`.
+
+For the example, this phase results in the DOM nodes being cloned into a `fragment`, ready to be inserted in the document, and a list of `Parts` that can be updated by the next phase to set the dynamic JavaScript values in that fragment.
+
+```js
+const instance = new TemplateInstance(preparedTemplate);
+const fragment = instance._clone();
+
+console.log(instance._$parts);
+/*
+[AttributePart, ChildPart, EventPart]
+*/
+```
+
+After the update phase this fragment is inserted into the DOM.
+
+The following sections cover in detail what happens when the `TemplateInstance` is created and then cloned:
+
+#### Create a `TemplateInstance`
```

Comment:
I'm not sure these two steps actually warrant being separate steps. It feels like `._clone()` really is more like part of the creation of the instance with updating the parts. The only reason it's not part of the constructor is to not hold reference to the fragment that's needed for creating the parts.

Overall, I feel the creation step is a "single" operation of creating the `TemplateInstance` which produces a document fragment to be inserted to the container, and the parts which hold references to the nodes within that fragment for future updates.

This template instance, with its parts, is then stored on the child part of the container, so that future render requests can pick it up and move on directly to the update phase.

Reply:
Agree! - Also applied your great suggestion to mention that this `TemplateInstance` is stored on the ChildPart such that it's possible to skip straight to `Update` in future renders.